File: checkpoint6/frontend/src/domain/trace.ts
/* ------------------------------------------------------------------ */
/*  Domain model                                                       */
/* ------------------------------------------------------------------ */

export interface TraceModel {
  jobId: string;
  status: string;
  currentBatchIndex: number;
  totalBatches: number;
  processedUnits: number;
  totalUnits: number;
  currentActivity?: string;
  recentEvents: TraceEventModel[];
  stats?: TraceStatsModel;
}

export interface TraceEventModel {
  id: string;
  jobId?: string;
  eventType: string;
  timestamp: string;
  severity?: 'info' | 'warn' | 'error' | 'debug';
  message?: string;
  data?: unknown;
  batchIndex?: number;
}

export interface TraceStatsModel {
  totalEvents: number;
  unitTranslated: number;
  retries: number;
  fallbacks: number;
  errors: number;
  failedUnits?: number;
  cacheHits?: number;
}

export interface TraceUnitModel {
  unitId: string;
  filePath: string;
  key: string;
  sourceText: string;
  translatedText: string;
  status: 'pending' | 'sent' | 'translated' | 'failed' | 'cached';
  errorMessage: string;
  batchIndex: number;
  updatedAt: string;
}

/* ------------------------------------------------------------------ */
/*  DTO → Domain mappers                                               */
/* ------------------------------------------------------------------ */

import type { TraceSnapshot, TraceEvent, TraceStats, TraceUnitResponse } from "../api/types";

export function mapTraceSnapshot(dto: TraceSnapshot): TraceModel {
  return {
    jobId: dto.job_id,
    status: dto.status,
    currentBatchIndex: dto.current_batch_index,
    totalBatches: dto.total_batches,
    processedUnits: dto.processed_units,
    totalUnits: dto.total_units,
    currentActivity: dto.current_activity || undefined,
    recentEvents: (dto.recent_events ?? []).map(e => mapTraceEvent(e)),
    stats: dto.stats ? mapTraceStats(dto.stats) : undefined,
  };
}

export function mapTraceEvent(dto: TraceEvent): TraceEventModel {
  return {
    id: dto.id,
    jobId: dto.job_id || undefined,
    eventType: dto.event_type,
    timestamp: dto.timestamp,
    severity: dto.severity || undefined,
    message: dto.message || undefined,
    data: dto.data ?? undefined,
    batchIndex: dto.batch_index ?? undefined,
  };
}

export function mapTraceStats(dto: TraceStats): TraceStatsModel {
  return {
    totalEvents: dto.total_events,
    unitTranslated: dto.unit_translated,
    retries: dto.retries,
    fallbacks: dto.fallbacks,
    errors: dto.errors,
    failedUnits: dto.failed_units ?? 0,
    cacheHits: dto.cache_hits ?? 0,
  };
}

export function mapTraceUnitResponse(dto: TraceUnitResponse): TraceUnitModel {
  return {
    unitId: dto.unit_id,
    filePath: dto.file_path,
    key: dto.key,
    sourceText: dto.source_text,
    translatedText: dto.translated_text,
    status: dto.status,
    errorMessage: dto.error_message,
    batchIndex: dto.batch_index,
    updatedAt: dto.updated_at,
  };
}

/* ------------------------------------------------------------------ */
/*  Helpers                                                            */
/* ------------------------------------------------------------------ */

export function isTraceActive(trace: TraceModel): boolean {
  return trace.status === "running" || trace.status === "paused";
}

export function getTraceStatus(trace: TraceModel): string {
  return trace.status;
}

export function getTraceEventLabel(eventType: string): string {
  const labels: Record<string, string> = {
    'job_started': 'Job started',
    'job_completed': 'Job completed',
    'job_failed': 'Job failed',
    'batch_started': 'Batch started',
    'batch_completed': 'Batch completed',
    'batch_failed': 'Batch failed',
    'retry_attempt': 'Retry',
    'fallback_started': 'Fallback started',
    'fallback_completed': 'Fallback completed',
    'prompt_built': 'Prompt built',
    'runtime_request_started': 'API request',
    'runtime_response_received': 'API response',
    'response_parsed': 'Response parsed',
    'unit_translated': 'Unit translated',
    'unit_failed': 'Unit failed',
    'cache_hit': 'Cache hit',
    'cache_miss': 'Cache miss',
  };
  return labels[eventType] || eventType;
}

export function getTraceEventBadgeClass(eventType: string): string {
  switch (eventType) {
    case 'job_started': case 'batch_started': case 'batch_completed':
      return 'badge-info';
    case 'retry_attempt': case 'batch_failed':
      return 'badge-warning';
    case 'job_failed': case 'unit_failed':
      return 'badge-error';
    case 'job_completed':
      return 'badge-success';
    case 'cache_hit':
      return 'badge-info';
    default:
      return 'badge-muted';
  }
}

export function getTraceCurrentActivity(trace: TraceModel): string {
  const activity = trace.currentActivity;
  if (!activity) return 'Processing...';
  return activity;
}

export function getTraceUnitBadgeClass(status: string): string {
  switch (status) {
    case 'translated': return 'badge-success';
    case 'failed':     return 'badge-error';
    case 'cached':     return 'badge-info';
    case 'sent':       return 'badge-warning';
    default:           return 'badge-muted';
  }
}

export function getTraceUnitStatusLabel(status: string): string {
  switch (status) {
    case 'translated': return 'Translated';
    case 'failed':     return 'Failed';
    case 'cached':     return 'Cached';
    case 'sent':       return 'Sent';
    default:           return 'Pending';
  }
}

/* ------------------------------------------------------------------ */
/*  Runtime status log helpers                                          */
/* ------------------------------------------------------------------ */

/** CSS badge class for the severity of a runtime event. */
export function getSeverityBadgeClass(severity?: string): string {
  switch (severity) {
    case 'error': return 'badge-error';
    case 'warn':  return 'badge-warning';
    case 'debug': return 'badge-muted';
    default:      return 'badge-info';
  }
}

/** Human-readable label for a severity level. */
export function getSeverityLabel(severity?: string): string {
  switch (severity) {
    case 'error': return 'ERROR';
    case 'warn':  return 'WARN';
    case 'debug': return 'DEBUG';
    default:      return 'INFO';
  }
}

/** Check whether a BATCH_COMPLETED event is clickable (has a detectable batch number). */
export function isClickableBatchEvent(ev: TraceEventModel): boolean {
  return ev.eventType === 'batch_completed' && extractBatchNo(ev) !== null;
}

/** Try to extract a batch number from a BATCH_COMPLETED event.
 *
 *  First checks the event-level ``batchIndex`` field (mapped from the
 *  backend ``TranslationTraceEvent.batch_index``), then falls back to
 *  scanning ``data`` for ``batch_index``, ``batch_no``, or ``batchIndex``
 *  (legacy backends that embed it in data).
 */
export function extractBatchNo(ev: TraceEventModel): number | null {
  if (ev.eventType !== 'batch_completed') return null;
  // Prefer the event-level batchIndex (mapped from backend model field)
  if (ev.batchIndex != null && typeof ev.batchIndex === 'number') return ev.batchIndex;
  // Fallback: scan data dict (legacy backends)
  const d = ev.data as Record<string, unknown> | undefined;
  if (d) {
    if (typeof d.batch_index === 'number') return d.batch_index;
    if (typeof d.batch_no === 'number') return d.batch_no;
    if (typeof d.batchIndex === 'number') return d.batchIndex;
  }
  return null;
}

/** @deprecated Use isClickableBatchEvent instead. */
export function isBatchCompletedEvent(ev: TraceEventModel): boolean {
  return isClickableBatchEvent(ev);
}

/** @deprecated Use extractBatchNo instead. */
export function getBatchNoFromEvent(ev: TraceEventModel): number | null {
  return extractBatchNo(ev);
}

/** Extract http_status from event data for runtime events. */
export function getHttpStatusFromEvent(ev: TraceEventModel): number | null {
  if (!ev.data || typeof ev.data !== 'object') return null;
  const d = ev.data as Record<string, unknown>;
  if (typeof d.http_status === 'number') return d.http_status;
  return null;
}

/** Extract retry_reason from event data. */
export function getRetryReasonFromEvent(ev: TraceEventModel): string | null {
  if (!ev.data || typeof ev.data !== 'object') return null;
  const d = ev.data as Record<string, unknown>;
  if (typeof d.retry_reason === 'string') return d.retry_reason;
  return null;
}

/** Get the event_type for runtime-event-specific badge coloring. */
export function getRuntimeEventBadgeClass(eventType: string): string {
  switch (eventType) {
    case 'runtime_request_started':   return 'badge-info';
    case 'runtime_response_received': return 'badge-info';
    case 'retry_attempt':             return 'badge-warning';
    case 'fallback_started':          return 'badge-warning';
    case 'fallback_completed':        return 'badge-info';
    case 'cache_hit':                 return 'badge-success';
    case 'cache_miss':                return 'badge-muted';
    default:                          return 'badge-muted';
  }
}

File: checkpoint6/src/translator_app/translation/trace.py
"""TranslationTraceService — structured event store for live debug / tracing.

Stores ``TranslationTraceEvent`` objects in-memory with an optional
``max_events`` cap to avoid unbounded memory growth.  All messages
and data dicts are sanitised via ``LogSanitizer`` before storage.

Can optionally accept a ``SecretsService`` to pass ``known_keys`` to the
sanitizer for deeper secret masking.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from translator_app.diagnostics.sanitizer import LogSanitizer
from translator_app.translation.trace_models import (
    TraceEventSeverity,
    TraceEventType,
    TraceUnitStatus,
    TraceUnitEntry,
    TranslationTraceEvent,
    TranslationTraceSnapshot,
    TranslationTraceStats,
)


_RUNTIME_EVENT_TYPES = frozenset({
    TraceEventType.RUNTIME_REQUEST_STARTED,
    TraceEventType.RUNTIME_RESPONSE_RECEIVED,
    TraceEventType.RETRY_ATTEMPT,
    TraceEventType.FALLBACK_STARTED,
    TraceEventType.FALLBACK_COMPLETED,
    TraceEventType.CACHE_HIT,
    TraceEventType.CACHE_MISS,
    TraceEventType.PROMPT_BUILT,
})


class TranslationTraceService:
    """In-memory trace event store with sanitization and max_events limit.

    Responsible for collecting structured trace events emitted by the
    translation pipeline (adapter, runtime, job execution) and providing
    query access for the live-debug API.
    """

    def __init__(
        self,
        max_events: int = 5000,
        max_units_per_job: int = 2000,
        secrets_service: Any = None,
        repo: Any = None,
    ):
        """
        Args:
            max_events: Maximum number of trace events to keep in memory.
                        Older events are dropped when the limit is exceeded.
            max_units_per_job: Maximum number of trace unit entries to keep
                               per job. Oldest entries are dropped when the
                               limit is exceeded.
            secrets_service: Optional SecretsService whose known key IDs
                             are passed to the LogSanitizer for deeper masking.
            repo: Optional TraceRepository for SQLite persistence (P0-06).
                  When provided, events and units are persisted and reloaded
                  on startup.
        """
        self._max_events = max_events
        self._max_units_per_job = max_units_per_job
        self._events: List[TranslationTraceEvent] = []
        self._trace_units: Dict[str, List[TraceUnitEntry]] = {}
        self._known_keys: List[str] = []
        self._repo = repo
        if secrets_service is not None:
            self._known_keys = list(getattr(secrets_service, "list_key_ids", lambda: [])())
        if self._repo is not None:
            self._load_from_db()

    # ------------------------------------------------------------------
    # Database persistence (P0-06)
    # ------------------------------------------------------------------

    def _load_from_db(self) -> None:
        """Load events and units from SQLite into the in-memory cache.

        Called on startup so that get_events / get_units return data
        from previous sessions.
        """
        if self._repo is None:
            return
        try:
            db_events = self._repo.get_all_events(limit=self._max_events)
            self._events = list(db_events)  # already ASC order
        except Exception:
            self._events = []

        try:
            for jid in self._repo.get_distinct_job_ids_with_units():
                units = self._repo.get_units_by_job(jid, limit=self._max_units_per_job)
                if units:
                    self._trace_units[jid] = units
        except Exception:
            self._trace_units = {}

    def _maybe_trim_db_events(self) -> None:
        """Trim oldest events in the DB beyond max_events limit."""
        if self._repo is None:
            return
        try:
            self._repo.trim_events(self._max_events)
        except Exception:
            pass

    def _maybe_trim_db_units(self, job_id: str) -> None:
        """Trim oldest units in the DB beyond max_units_per_job limit."""
        if self._repo is None:
            return
        try:
            self._repo.trim_units(job_id, self._max_units_per_job)
        except Exception:
            pass

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def add_event(
        self,
        event_type: TraceEventType,
        job_id: str = "",
        task_id: str = "",
        batch_index: int = 0,
        unit_ids: Optional[List[str]] = None,
        provider: str = "",
        model: str = "",
        src_lang: str = "",
        dst_lang: str = "",
        severity: TraceEventSeverity = TraceEventSeverity.INFO,
        message: str = "",
        data: Optional[Dict[str, Any]] = None,
        diagnostics: Optional[Dict[str, Any]] = None,
    ) -> TranslationTraceEvent:
        """Create, sanitise, store and return a new trace event.

        All string messages and data dicts are sanitised via
        ``LogSanitizer`` before being stored.
        """
        sanitised_message = LogSanitizer.sanitize(message, known_keys=self._known_keys)
        sanitised_data = LogSanitizer.sanitize_context(data, known_keys=self._known_keys) if data else None
        sanitised_diag = LogSanitizer.sanitize_context(diagnostics, known_keys=self._known_keys) if diagnostics else None

        event = TranslationTraceEvent(
            job_id=job_id,
            task_id=task_id,
            batch_index=batch_index,
            unit_ids=unit_ids or [],
            event_type=event_type,
            timestamp=datetime.utcnow(),
            provider=provider,
            model=model,
            src_lang=src_lang,
            dst_lang=dst_lang,
            severity=severity,
            message=sanitised_message,
            data=sanitised_data,
            diagnostics=sanitised_diag,
        )

        self._events.append(event)

        # P0-06: persist to SQLite if repo is available
        if self._repo is not None:
            try:
                self._repo.save_event(event)
            except Exception:
                pass

        # Trim oldest events when over limit
        if len(self._events) > self._max_events:
            self._events = self._events[-self._max_events:]
            self._maybe_trim_db_events()

        return event

    def get_events(
        self,
        job_id: str,
        limit: int = 100,
        event_type: Optional[TraceEventType] = None,
    ) -> List[TranslationTraceEvent]:
        """Return events for a job, optionally filtered by event type.

        Results are ordered newest-first.
        """
        filtered = [e for e in self._events if e.job_id == job_id]

        if event_type is not None:
            filtered = [e for e in filtered if e.event_type == event_type]

        # Newest first
        filtered.sort(key=lambda e: e.timestamp, reverse=True)
        return filtered[:limit]

    def get_runtime_events(
        self,
        job_id: str,
        limit: int = 100,
    ) -> List[TranslationTraceEvent]:
        """Return runtime-status-related events for a job, newest-first.

        Includes: RUNTIME_REQUEST_STARTED, RUNTIME_RESPONSE_RECEIVED,
        RETRY_ATTEMPT, FALLBACK_STARTED, FALLBACK_COMPLETED,
        CACHE_HIT, CACHE_MISS, PROMPT_BUILT.
        """
        filtered = [
            e for e in self._events
            if e.job_id == job_id and e.event_type in _RUNTIME_EVENT_TYPES
        ]
        filtered.sort(key=lambda e: e.timestamp, reverse=True)
        return filtered[:limit]

    def get_recent(self, limit: int = 100) -> List[TranslationTraceEvent]:
        """Return the most recent events across all jobs."""
        sorted_events = sorted(self._events, key=lambda e: e.timestamp, reverse=True)
        return sorted_events[:limit]

    def get_snapshot(self, job_id: str) -> TranslationTraceSnapshot:
        """Build a point-in-time snapshot for a job.

        Includes recent events, aggregated stats, and current batch info.
        """
        job_events = [e for e in self._events if e.job_id == job_id]
        job_events_sorted = sorted(job_events, key=lambda e: e.timestamp, reverse=True)

        # Determine current batch from the most recent batch_started event
        current_batch = 0
        for ev in job_events_sorted:
            if ev.event_type == TraceEventType.BATCH_STARTED:
                current_batch = ev.batch_index
                break

        # Count total batches from batch_completed + batch_failed events
        completed_batches = sum(
            1 for e in job_events
            if e.event_type in (TraceEventType.BATCH_COMPLETED, TraceEventType.BATCH_FAILED)
        )
        total_batches = max(completed_batches, current_batch)

        # Collect current unit IDs from the most recent batch_started
        current_units: List[str] = []
        for ev in job_events_sorted:
            if ev.event_type == TraceEventType.BATCH_STARTED and ev.unit_ids:
                current_units = ev.unit_ids
                break

        # Determine status
        status = "unknown"
        if any(e.event_type == TraceEventType.JOB_FAILED for e in job_events):
            status = "failed"
        elif any(e.event_type == TraceEventType.JOB_COMPLETED for e in job_events):
            status = "completed"
        elif any(e.event_type == TraceEventType.JOB_STARTED for e in job_events):
            status = "running"

        stats = self.get_stats(job_id)

        # Determine current activity from the most recent events
        current_activity = self._compute_current_activity(job_events_sorted, status, current_batch, total_batches)

        return TranslationTraceSnapshot(
            job_id=job_id,
            status=status,
            current_batch_index=current_batch,
            total_batches=total_batches,
            current_units=current_units,
            recent_events=job_events_sorted[:10],
            stats=stats,
            current_activity=current_activity,
        )

    def get_stats(self, job_id: str) -> TranslationTraceStats:
        """Compute aggregated statistics for a job."""
        job_events = [e for e in self._events if e.job_id == job_id]

        # Collect latencies from batch_completed diagnostics
        latencies: List[float] = []
        for ev in job_events:
            if ev.event_type == TraceEventType.BATCH_COMPLETED and ev.diagnostics:
                latency = ev.diagnostics.get("latency_ms")
                if latency is not None:
                    latencies.append(float(latency))

        avg_latency = (sum(latencies) / len(latencies)) if latencies else 0.0

        return TranslationTraceStats(
            total_events=len(job_events),
            batches_started=sum(
                1 for e in job_events if e.event_type == TraceEventType.BATCH_STARTED
            ),
            batches_completed=sum(
                1 for e in job_events if e.event_type == TraceEventType.BATCH_COMPLETED
            ),
            retries=sum(
                1 for e in job_events if e.event_type == TraceEventType.RETRY_ATTEMPT
            ),
            fallbacks=sum(
                1 for e in job_events
                if e.event_type in (TraceEventType.FALLBACK_STARTED, TraceEventType.FALLBACK_COMPLETED)
            ),
            cache_hits=sum(
                1 for e in job_events if e.event_type == TraceEventType.CACHE_HIT
            ),
            cache_misses=sum(
                1 for e in job_events if e.event_type == TraceEventType.CACHE_MISS
            ),
            failed_units=sum(
                1 for e in job_events if e.event_type == TraceEventType.UNIT_FAILED
            ),
            avg_latency_ms=avg_latency,
        )

    @staticmethod
    def _compute_current_activity(
        events_sorted: List[TranslationTraceEvent],
        status: str,
        current_batch: int,
        total_batches: int,
    ) -> str:
        """Build a human-readable current activity string from events."""
        if not events_sorted:
            return ""

        most_recent = events_sorted[0].event_type if events_sorted else None

        if status == "completed":
            return "Completed"
        if status == "failed":
            return "Failed"

        if most_recent in (TraceEventType.RETRY_ATTEMPT,):
            if current_batch and total_batches:
                return f"Retrying batch {current_batch}/{total_batches}"
            return "Retrying batch"

        if current_batch and total_batches:
            if most_recent in (TraceEventType.FALLBACK_STARTED, TraceEventType.FALLBACK_COMPLETED):
                return f"Falling back to single mode (batch {current_batch}/{total_batches})"
            return f"Translating batch {current_batch}/{total_batches}"

        if most_recent == TraceEventType.JOB_STARTED:
            return "Starting translation"

        return ""

    # ------------------------------------------------------------------
    # Trace unit management (live source/translation monitoring)
    # ------------------------------------------------------------------

    def add_unit(self, entry: TraceUnitEntry) -> None:
        """Store or update a trace unit entry for a job.

        Sanitizes source and translated text before storing.
        Respects max_units_per_job cap by trimming oldest entries.
        """
        sanitized_source = LogSanitizer.sanitize(entry.source_text, known_keys=self._known_keys)
        sanitized_translated = LogSanitizer.sanitize(entry.translated_text, known_keys=self._known_keys)
        sanitized_error = LogSanitizer.sanitize(entry.error_message, known_keys=self._known_keys)

        sanitized_entry = entry.model_copy(update={
            "source_text": sanitized_source,
            "translated_text": sanitized_translated,
            "error_message": sanitized_error,
            "updated_at": datetime.utcnow(),
        })

        job_units = self._trace_units.setdefault(entry.job_id, [])
        # Replace existing entry for the same unit_id, or append new one
        for i, existing in enumerate(job_units):
            if existing.unit_id == entry.unit_id:
                job_units[i] = sanitized_entry
                break
        else:
            job_units.append(sanitized_entry)

        # P0-06: persist to SQLite if repo is available
        if self._repo is not None:
            try:
                self._repo.save_unit(sanitized_entry)
            except Exception:
                pass

        # Trim oldest entries when over limit
        if len(job_units) > self._max_units_per_job:
            self._trace_units[entry.job_id] = job_units[-self._max_units_per_job:]
            self._maybe_trim_db_units(entry.job_id)

    def add_unit_batch(
        self,
        job_id: str,
        entries: List[TraceUnitEntry],
    ) -> None:
        """Store multiple trace unit entries for a job."""
        for entry in entries:
            entry.job_id = job_id
            self.add_unit(entry)

    def get_units(
        self,
        job_id: str,
        limit: int = 50,
    ) -> List[TraceUnitEntry]:
        """Return trace unit entries for a job, newest-first, capped by limit."""
        job_units = self._trace_units.get(job_id, [])
        # Return newest first
        sorted_units = sorted(job_units, key=lambda u: u.updated_at, reverse=True)
        return sorted_units[:limit]

    def get_units_by_batch(
        self,
        job_id: str,
        batch_index: int,
        limit: int = 500,
    ) -> List[TraceUnitEntry]:
        """Return trace unit entries for a specific batch, newest-first."""
        job_units = self._trace_units.get(job_id, [])
        filtered = [u for u in job_units if u.batch_index == batch_index]
        # Return newest first
        filtered.sort(key=lambda u: u.updated_at, reverse=True)
        return filtered[:limit]

    def clear(self, job_id: Optional[str] = None) -> None:
        """Clear events and trace units for a specific job or all.

        Also removes persisted data from SQLite if a repo is available.
        """
        if job_id is None:
            self._events.clear()
            self._trace_units.clear()
            if self._repo is not None:
                try:
                    self._repo.delete_all_events()
                    self._repo.delete_all_units()
                except Exception:
                    pass
        else:
            self._events = [e for e in self._events if e.job_id != job_id]
            self._trace_units.pop(job_id, None)
            if self._repo is not None:
                try:
                    self._repo.delete_events_by_job(job_id)
                    self._repo.delete_units_by_job(job_id)
                except Exception:
                    pass

    @property
    def event_count(self) -> int:
        """Total number of stored events."""
        return len(self._events)

    @property
    def max_events(self) -> int:
        """Maximum number of events allowed."""
        return self._max_events
